### PR TITLE
Enforce rubocop-rspec for project

### PR DESCRIPTION
**Pull request by backus**
20% snark PR responding to @Drenmi's comment on https://github.com/bbatsov/rubocop/issues/4653. 80% serious

This PR **does not** ship rubocop-rspec with RuboCop. All it does is enforce rubocop-rspec's cops for the rubocop project itself.

I've applied each autocorrect in a separate commit so people can pick and chose what they want to keep. Harder to update stuff is in .rubocop_todo.yml. I've also updated some stuff in past PRs for fun.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [ ] Squashed related commits together.
* [ ] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: .rubocop_todo.yml [hunk 1]
```diff
@@ -1,21 +1,29 @@
...
 # The point is for the user to remove these configuration records
 # one by one as the offenses are removed from the code base.
 # Note that changes in the inspected code, or installation of new
 # versions of RuboCop, may require this file to be generated again.
 
+# Offense count: 1
+# Cop supports --auto-correct.
+# Configuration parameters: Include, TreatCommentsAsGroupSeparators.
+# Include: **/Gemfile, **/gems.rb
+Bundler/OrderedGems:
```
Comment (bbatsov) [inline .rubocop_todo.yml]: This seems unrelated to the scope of the PR.

**Comment by bbatsov** [comment]
I see we have a problem with RuboCop still supporting Ruby 2.0, unlike `rubocop-rspec`. There's also some JRuby failure that might be related (looking at the stacktrace and seeing it's about a missing method).

**Comment by Drenmi** [comment]
Nice! I'm all for it. The tests are the least consistent part about the code base now, and I think this would help a lot.

---

I intentionally did not go with the `be_*` assertion style for node extensions, as many of the predicates are about the node's composition, rather than behaviour. E.g. a `yield` node is not a splat argument, but it can have a splat argument.

This could be amended with a much more elaborate naming scheme I guess, e.g. `SendNode#parameterized?` instead of `SendNode#arguments?`, but I don't think most people would intuitively look for that name. 🙂 

What's your thought on this @backus?

File: .rubocop_todo.yml (hunk 1 above)
Comment (backus) [reply .rubocop_todo.yml]: Yeah that was a mistake. Fixing

**Comment by backus** [comment]
@Drenmi I expected the `be_*` matchers to be one of the discussions this PR would bring up. Luckily, @pocke contributed this cop to rubocop-rspec and I think it is maybe the most impressive auto-correct it ships now. Checkout the configuration: 

https://github.com/backus/rubocop-rspec/blob/e3a5413c7a06e7dc6ccc0225fac94b03a2687a19/config/default.yml#L307-L315

I'll update this PR to use the strict and explicit style.

**Comment by backus** [comment]
@bbatsov a jruby fix should be released soon. For ruby 2.0 best option is probably to just not run rubocop-rspec for that build.

**Comment by Drenmi** [comment]
> I expected the be_* matchers to be one of the discussions this PR would bring up.

I am not at all dogmatic about this. I really like the `be_*` matchers, because it encourages people to think about their naming. I think AST nodes are a rare exception, since it makes sense to have predicates describing the object's composition with regards to child nodes. 🙂

**Comment by bbatsov** [comment]
I never had a strong opinion about this myself, so I'm fine with whatever as long as we're (mostly consistent). 

@backus Out of curiosity - why 2.1 and not say 2.2. I'm just wondering if it was because of some language feature or dependency or because you wanted to supported only supported Rubies. I myself plan to kill the support for 2.0 and 2.1 soon, because they haven't been supported for a while and I doubt many people still use them given the trivial upgrade path.

**Comment by backus** [comment]
> Out of curiosity - why 2.1 and not say 2.2. I'm just wondering if it was because of some language feature or dependency or because you wanted to supported only supported Rubies

I removed both and people whined about 2.2 being dropped so I added it back: https://github.com/backus/rubocop-rspec/pull/131

Actually a pretty effective way to make the decision lol.

**Comment by Drenmi** [comment]
Following Ruby's own EOL seems reasonable. Enabling people to defer upgrading their dependencies is a disservice to all. Friends don't let friends stay on old versions. 😉

**Comment by bbatsov** [comment]
> I removed both and people whined about 2.2 being dropped so I added it back: backus/rubocop-rspec#131

That's really interesting - with the trivial upgrade path from 2.1 to 2.2+ I can't imagine what's the reason someone would stick with it.

**Comment by bbatsov** [comment]
I see what happened - @backus dropped support for 2.1 before it reached its EOL and that's why some people complained. I believe right now it's perfectly fine to drop support for 2.1.

**Comment by backus** [comment]
> I believe right now it's perfectly fine to drop support for 2.1.

Hurray. Less work for me 😆

**Comment by bbatsov** [comment]
@backus Please, rebase on top of the current `master` so I can merge this.

We dropped support for 2.0 in a PR I just merged, so we'd be on the same page for now and the build will not be an issue.

**Comment by bbatsov** [comment]
@backus ping :-)

**Comment by backus** [comment]
@bbatsov might require a trickier update given that rubocop-rspec isn't using keyword args right now.

**Comment by backus** [comment]
Yeah this change messes with at least one of the cops. Might have to figure out another tricky rubocop-rspec release

**Comment by bbatsov** [comment]
> might require a trickier update given that rubocop-rspec isn't using keyword args right now.

But we kept the old API for offenses as well, so this so be good. There will just be a bunch of deprecation messages.

**Comment by bbatsov** [comment]
The build seems to be failing.

**Comment by backus** [comment]
This change https://github.com/bbatsov/rubocop/commit/d7cddd225 breaks the compatibility with https://github.com/backus/rubocop-rspec/blob/02d07af20d6461f624f5cec4097280c1eaf99916/lib/rubocop/cop/rspec/describe_class.rb#L25-L27 so I'll need to do a special release or something

**Comment by backus** [comment]
Hurray! Build is finally green @bbatsov. Thanks for the well timed release I pushed out a compatibility release for rubocop-rspec now

**Comment by bbatsov** [comment]
@backus We're one merge conflict away from finally wrapping this. :)

**Comment by bbatsov** [comment]
👍

**Comment by Drenmi** [comment]
🎉 🎉 🎉
